sympy/combinatorics/pc_groups.py

```diff
@@ -0,0 +1,283 @@
+        for key, value in self.pc_relators.items():
+            if len(key.array_form) == 1:
+                power_relators[key] = value
+            else:
+                conjugate_relators[key] = value
+        return power_relators, conjugate_relators
+
+    def subword_index(self, w):
+        """
+        Returns the start and ending index of a given
+        subword in a word.
+
+        Examples
+        ========
+        >>> from sympy.combinatorics.pc_groups import Collector
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> F, x1, x2 = free_group("x1, x2")
+        >>> pc_relators = {x1**2 : 1, x1*x2*x1**-1 : x2**-1, x1**-1*x2*x1 : x2**-1}
+        >>> word = x2**2*x1**7
+        >>> relative_order = {x1: 2, x2: 3}
+        >>> collector = Collector(pc_relators, word, relative_order)
+        >>> w = x2**2*x1
+        >>> collector.subword_index(w)
+        (0, 3)
+        >>> w = x1**7
+        >>> collector.subword_index(w)
+        (2, 9)
+
+        """
+        low = -1
+        high = -1
+        for i in range(len(self.word)-len(w)+1):
+            if self.word.subword(i, i+len(w)) == w:
+                low = i
+                high = i+len(w)
+                break
+        if low == high == -1:
+            return -1, -1
+        return low, high
+
+    def map_relation(self, w):
+        """
+        Examples
+        ========
+        >>> from sympy.combinatorics.pc_groups import Collector
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> F, x0, x1, x2 = free_group("x0, x1, x2")
+        >>> pc_relators = {x0**-1*x1*x0: x1**2, x1**-1*x2*x1:x2, x0**-1*x2*x0:x2*x1}
+        >>> relative_order = {x0: 2, x1: 2, x2: 3}
+        >>> word = x2*x1*x0
+        >>> collector = Collector(pc_relators, word, relative_order)
+        >>> w = x2*x1
+        >>> collector.map_relation(w)
+        x2
+        >>> w = x1*x0
+        >>> collector.map_relation(w)
+        x1**2
+
+        """
+        group = w.group
+        gens = list(sorted(w.contains_generators()))
+        key = gens[0]**-1*gens[1]*gens[0]
+        return self.pc_relators[key]
+
+
+def collected_word(pc_relators, word, relative_order):
+    """
+    Examples
+    ========
+    >>> from sympy.combinatorics.pc_groups import collected_word
+    >>> from sympy.combinatorics.pc_groups import Collector
+    >>> from sympy.combinatorics.free_groups import free_group
+    >>> F, x1, x2 = free_group("x1, x2")
+    >>> pc_relators = {x1**2 : (), x1*x2*x1**-1 : x2**-1, x1**-1*x2*x1 : x2**-1}
+    >>> word = x2**2*x1**7
+    >>> relative_order = {x1: 2, x2: 3}
+    >>> collected_word(pc_relators, word, relative_order)
+    x1*x2**-2
+
+    """
+    collector = Collector(pc_relators, word, relative_order)
+    power_relators, conjugate_relators = collector.relations()
+    group = collector.group
+    flag = 0
+    while True:
+        uncollected_subwords = collector.minimal_uncollected_subwords()
```

Comment:
It would suffice to get one minimal uncollected subword at a time. If there are several, those will be handled on the following rounds of the True loop. It would probably be enough to have the word in array form and its position. Its type and length can be easily found here.

Reply:
Do you mean just processing a single uncollected subword at a time that is, removing the loop below this?

Reply:
Yes. The True-loop alone should suffice.